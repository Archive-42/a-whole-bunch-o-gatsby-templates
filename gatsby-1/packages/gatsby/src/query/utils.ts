import { Path } from "graphql/jsutils/Path"

export const indentString = (string: string): string =>
  string.replace(/\n/g, `\n  `)

export const formatErrorDetails = (errorDetails: Map<string, any>): string =>
  Array.from(errorDetails.entries())
    .map(
      ([name, details]) => `${name}:
  ${indentString(details.toString())}`
    )
    .join(`\n`)

export function pathToArray(path: Path | undefined): Array<string | number> {
  const flattened: Array<string | number> = []
  let curr: Path | undefined = path
  while (curr) {
    flattened.push(curr.key)
    curr = curr.prev
  }
  return flattened.reverse()
}
