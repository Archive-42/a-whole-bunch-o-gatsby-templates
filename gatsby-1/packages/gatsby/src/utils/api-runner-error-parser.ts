import { IMatch } from "../types"

const errorParser = ({ err }: { err: unknown }): IMatch => {
  const handlers = [
    {
      regex: /(.+) is not defined/m,
      cb: (match: RegExpMatchArray): IMatch => {
        return {
          id: `11330`,
          context: { sourceMessage: match[0], arg: match[1] },
        }
      },
    },
  ]

  let structured: IMatch | undefined
  let errorMessage: string | undefined

  // try to handle as many type of err as possible.
  // the err might come from a plugin so we don't
  // know what we are getting
  if (Array.isArray(err)) {
    err = err[0]
  }
  if (err instanceof Error) {
    errorMessage = err.message
  }
  if (typeof err === `string`) {
    errorMessage = err
  }

  for (const { regex, cb } of handlers) {
    const matched = errorMessage?.match(regex)
    if (matched) {
      structured = {
        ...cb(matched),
      }
      break
    }
  }

  // if we haven't found any known error
  if (!structured) {
    return {
      id: `11321`,
      context: { sourceMessage: errorMessage || `` },
      error: err instanceof Error ? err : undefined,
    }
  }

  return structured
}

export default errorParser
