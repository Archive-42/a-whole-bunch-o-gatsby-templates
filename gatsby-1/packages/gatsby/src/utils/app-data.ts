import fs from "fs-extra"
import path from "path"

const APP_DATA_JSON = `app-data.json`

export const write = (publicDir: string, hash: string): Promise<void> =>
  fs.outputJson(path.join(publicDir, `page-data`, APP_DATA_JSON), {
    webpackCompilationHash: hash,
  })

export const exists = (publicDir: string): boolean =>
  fs.pathExistsSync(path.join(publicDir, `page-data`, APP_DATA_JSON))
