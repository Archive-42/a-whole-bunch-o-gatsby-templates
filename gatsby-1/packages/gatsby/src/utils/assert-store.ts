import { Store } from "redux"
import reporter from "gatsby-cli/lib/reporter"

export function assertStore(store?: Store): asserts store {
  if (!store) {
    reporter.panic(`Could not find Redux store`)
  }
}
