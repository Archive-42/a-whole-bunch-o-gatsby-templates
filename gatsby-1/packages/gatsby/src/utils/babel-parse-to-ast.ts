import { parse, ParserOptions } from "@babel/parser"
import { File } from "@babel/types"

const PARSER_OPTIONS: ParserOptions = {
  allowImportExportEverywhere: true,
  allowReturnOutsideFunction: true,
  allowSuperOutsideMethod: true,
  sourceType: `unambiguous`,
  plugins: [
    `jsx`,
    `flow`,
    `doExpressions`,
    `objectRestSpread`,
    [
      `decorators`,
      {
        decoratorsBeforeExport: true,
      },
    ],
    `classProperties`,
    `classPrivateProperties`,
    `classPrivateMethods`,
    `exportDefaultFrom`,
    `exportNamespaceFrom`,
    `asyncGenerators`,
    `functionBind`,
    `functionSent`,
    `dynamicImport`,
    `numericSeparator`,
    `optionalChaining`,
    `importMeta`,
    `bigInt`,
    `optionalCatchBinding`,
    `throwExpressions`,
    [
      `pipelineOperator`,
      {
        proposal: `minimal`,
      },
    ],
    `nullishCoalescingOperator`,
  ],
}

export function getBabelParserOptions(filePath: string): ParserOptions {
  // Flow and TypeScript plugins can't be enabled simultaneously
  if (/\.tsx?/.test(filePath)) {
    const { plugins } = PARSER_OPTIONS
    return {
      ...PARSER_OPTIONS,
      plugins: plugins!.map(plugin =>
        plugin === `flow` ? `typescript` : plugin
      ),
    }
  }
  return PARSER_OPTIONS
}

export function babelParseToAst(contents: string, filePath: string): File {
  return parse(contents, getBabelParserOptions(filePath))
}
