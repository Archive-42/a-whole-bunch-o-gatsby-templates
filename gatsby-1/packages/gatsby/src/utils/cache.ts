import manager, {
  Store,
  StoreConfig,
  CachingConfig,
  MultiCache,
} from "cache-manager"
import fs from "fs-extra"
import * as fsStore from "../cache/cache-fs"
import path from "path"

const MAX_CACHE_SIZE = 250
const TTL = Number.MAX_SAFE_INTEGER

interface ICacheProperties {
  name?: string
  store?: Store
}

export default class GatsbyCache {
  public name: string
  public store: Store
  public directory: string
  public cache?: MultiCache

  // @ts-ignore - set & get types are missing from fsStore?
  constructor({ name = `db`, store = fsStore }: ICacheProperties = {}) {
    this.name = name
    this.store = store
    this.directory = path.join(process.cwd(), `.cache/caches/${name}`)
  }

  init(): GatsbyCache {
    fs.ensureDirSync(this.directory)

    const configs: Array<StoreConfig> = [
      {
        store: `memory`,
        max: MAX_CACHE_SIZE,
        ttl: TTL,
      },
      {
        store: this.store,
        ttl: TTL,
        options: {
          path: this.directory,
          ttl: TTL,
        },
      },
    ]

    const caches = configs.map(cache => manager.caching(cache))

    this.cache = manager.multiCaching(caches)

    return this
  }

  async get<T = unknown>(key): Promise<T | undefined> {
    return new Promise(resolve => {
      if (!this.cache) {
        throw new Error(
          `GatsbyCache wasn't initialised yet, please run the init method first`
        )
      }
      this.cache.get<T>(key, (err, res) => {
        resolve(err ? undefined : res)
      })
    })
  }

  async set<T>(
    key: string,
    value: T,
    args: CachingConfig = { ttl: TTL }
  ): Promise<T | undefined> {
    return new Promise(resolve => {
      if (!this.cache) {
        throw new Error(
          `GatsbyCache wasn't initialised yet, please run the init method first`
        )
      }
      this.cache.set(key, value, args, err => {
        resolve(err ? undefined : value)
      })
    })
  }
}
