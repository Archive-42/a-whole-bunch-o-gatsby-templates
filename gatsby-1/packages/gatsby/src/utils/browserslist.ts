import path from "path"
import browserslist from "browserslist/node"
import query from "browserslist"

const installedGatsbyVersion = (directory: string): number | undefined => {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { version } = require(path.join(
      directory,
      `node_modules`,
      `gatsby`,
      `package.json`
    ))
    return parseInt(version.split(`.`)[0], 10)
  } catch (e) {
    return undefined
  }
}

export const getBrowsersList = (directory: string): Array<string> => {
  const fallback =
    installedGatsbyVersion(directory) === 1
      ? [`>1%`, `last 2 versions`, `IE >= 9`]
      : [`>0.25%`, `not dead`]

  const config = browserslist.findConfig(directory)

  if (config && config.defaults) {
    return config.defaults
  }

  return fallback
}

export const hasES6ModuleSupport = (directory: string): boolean => {
  const browserslist = getBrowsersList(directory)
  return query(browserslist + `, not supports es6-module`).length === 0
}
