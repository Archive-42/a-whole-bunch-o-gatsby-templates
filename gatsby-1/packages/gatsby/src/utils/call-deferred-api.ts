import { assertStore } from "./assert-store"
import { Store } from "redux"
import { IMutationAction } from "../services"
import { actions } from "../redux/actions"
import reporter from "gatsby-cli/lib/reporter"
/**
 * These are the deferred redux actions sent from api-runner-node
 * They may include a `resolve` prop (if they are createNode actions).
 * If so, we resolve the promise when we're done
 */
export const callRealApi = (event: IMutationAction, store?: Store): void => {
  assertStore(store)
  const { type, payload, resolve } = event
  if (type in actions) {
    // If this is a createNode action then this will be a thunk.
    // No worries, we just dispatch it like any other
    const action = actions[type](...payload)
    const result = store.dispatch(action)
    // Somebody may be waiting for this
    if (resolve) {
      resolve(result)
    }
  } else {
    reporter.log(`Could not dispatch unknown action "${type}`)
  }
}
