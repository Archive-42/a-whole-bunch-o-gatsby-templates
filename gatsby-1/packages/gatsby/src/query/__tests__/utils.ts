import { indentString, formatErrorDetails } from "../utils"

describe(`Text formatting `, () => {
  it(`indent string correctly`, () => {
    expect(
      indentString(
        `  Line 1
Line 2
  - Line 3`
      )
    ).toMatchSnapshot()
  })

  it(`format error details correctly`, () => {
    const testErrors = new Map()
    testErrors.set(`Field`, `One line error`)
    testErrors.set(
      `Bar`,
      `Three
line
error`
    )
    expect(formatErrorDetails(testErrors)).toMatchSnapshot()
  })
})
