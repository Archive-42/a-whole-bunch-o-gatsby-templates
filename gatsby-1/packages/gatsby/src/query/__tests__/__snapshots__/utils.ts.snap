// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text formatting  format error details correctly 1`] = `
"Field:
  One line error
Bar:
  Three
  line
  error"
`;

exports[`Text formatting  indent string correctly 1`] = `
"  Line 1
  Line 2
    - Line 3"
`;
